### DISCUSSION TITLE
Maybe bug: Creating diamond diagram using NewType

### DISCUSSION THREAD

**Discussion by Jacob-Stevens-Haas** [microsoft/pyright#5524] [General]
I tried to use `NewType` to create a diamond-inheritance subtype, similar to the following 

```python
Foo = NewType("Foo", int | str)
```

As mentioned in #1397, that's verboten.  And indeed, pyright/pylance say `Expected class as second argument to NewType`.  But why is that the case?  The second argument to `NewType` doesn't receive much detail in [the docs](https://docs.python.org/3/library/typing.html#newtype) ([also](https://docs.python.org/3/library/typing.html#typing.NewType)), other than referring to it as a type.  And [PEP 483](https://peps.python.org/pep-0483/#types-vs-classes) says 

> `Union[str, int]` is a type but not a proper class

(although perhaps typing docs mangles the distinction between classes and types with "treat the new type as if it were a subclass [not subtype?] of the original type")

Am I using the wrong tool for the job?  Is there not currently a way to create a diamond-diagram subtype for type checking without creating an actual subclass?

**Comment by erictraut** [comment]
When you create a `NewType`, the resulting type is supposed to act as though it's a class that derives from the specified type. That means the second argument must be a class. It can't be a union — just like you can't use a union as a base class in a `class` statement.

```python
class Foo(int | str): ... # Illegal!
```

Keep in mind that unions are not ordered, so `int | str` is the same type as `str | int`. By comparison, when you use multiple inheritance, the order of the base classes is significant because it affects the mro.

```python
# These are very different classes!
class Sub1(int, str): ...
class Sub2(str, int): ...
```

I don't understand what you're trying to achieve by creating "a diamond-diamond subtype for type checking without creating an actual subclass". There's potentially another way to accomplish your goal, but you'd have to explain more for me to offer a suggestion.

**Reply by Jacob-Stevens-Haas** [reply]
Sorry, that was a typo: ~diamond-diamond~ diamond-inheritance

I was trying to annotate functions that accept either `numpy.ndarray` or `scipy.sparse.spmatrix`.  So I had been using a simple alias to a union.  However, in addition to making function signatures clearer, I also want to catch and prevent errors due to mathematically incorrect usage with type checking.  Something like:

```python
StateVector = NewType("StateVector", np.ndarray | sparse.spmatrix)
InequalityConstraint = NewType("InequalityConstraint", np.ndarray | sparse.spmatrix)

def solve_optimization(x0: StateVector, lower_bound: InequalityConstraint) -> StateVector:
    """Optimize something, given x >= lower_bound"""
```
I'd like to have a type checker flag as an error if someone passes in two lower bounds, rather than a state vector.  Or if they use an `InequalityConstraint` in a routine that's designed for `EqualityConstraint`, etc.  

This is a simple example, but these kinds of matrices are often generated in one part of a library and passed around to routines in other parts.  I'm hoping that by formally restricting the types that algorithms accept, I can catch mistakes that would otherwise require careful numerical analysis to detect.  As the docs say

>  "This [NewType] is useful in helping catch logical errors"

Sorry for the repeated, rapid edits.

**Reply by erictraut** [reply]
A type created with `NewType` requires caller to explicitly construct that new value with that type — effectively wrapping their original ndarray or spmatrix in that new type. For this reason, `NewType` is rarely used for types in public library interfaces.

But if you're willing to require that your callers construct new "wrapper" values, you could simply define classes called `StateVector` and `InequalityConstraint`. These classes could act as lightweight wrappers around the array or matrix and provide strong typing. 

```python
@dataclass
class StateVector:
    data: ndarray | spmatrix

@dataclass
class InequalityConstraint:
    data: ndarray | spmatrix
```

You would need to change the implementation to access the attribute, but this would be easy to do.

```python
def solve_optimization(
    x0: StateVector, lower_bound: InequalityConstraint
) -> StateVector:
    x0_data = x0.data
    lb_data = lower_bound.data
    ...
```

**Reply by Jacob-Stevens-Haas** [reply]
Ah yes, I'm familiar that users would need to explicitly construct - but the user in this case is me in a few months after i'm working on a different part of the code, or close colleagues working on the same internal-use package, or inside private modules in a public package.  And thanks - lightweight wrappers could be useful in cases.

I wanted to return and ask about something you said:

> When you create a `NewType`, the resulting type is supposed to act as though it's a class that derives from the specified type.

I'm confused by what it means for a type to "act" as anything.  If I understand correctly (and I'm sure I don't), types in type checking only "act" in the sense that they obey a subtype relation.  At runtime, the object acts as it is, since it is unchanged.  Which is why I'm confused by why MRO would matter.

**Reply by erictraut** [reply]
By "act as", I mean that you can use all of the same methods and attributes that are available for the class.

MRO would matter if `NewType` were allowed to be created with more than one class, as you were attempting to do. When multiple classes are involved — especially when those classes both implement methods or attributes with the same name, MRO determines which class overrides and which is overridden.

**Reply by Jacob-Stevens-Haas** [reply]
Thanks for the follow up, Eric!  I understand why MRO is important for classes, but not for types.  In `Derived = NewType("Derived", tp)`, the `Derived` object is of class `NewType`, with no relation to the MRO of `tp`.  `Derived.__call__` is the identity function.  `tp` is assigned to`Derived.__supertype__`.  `NewType` defines an `__mro_entries__` method, but only for simplifying errors if someone accidentally subclasses `NewType`.  So I'm wondering if the prohibition of `NewType(name, Union[int, str])` is merely a consequence of wanting to prevent `class Foo(NewType)`.  ([source](https://github.com/python/cpython/blob/38553442c88c331d91562b77f023345e0b33b5f9/Lib/typing.py#L39))

I guess I'm asking, is the restriction of `tp` to being a class in `NewType(name, tp)` done for implementation reasons, even though the type theory concept is valid?  

Or am I misunderstanding the association of python typing syntax to typing theory in PEP 483?  I had believed the role of `NewType` was to declare an arbitrary subset (subtype) of a larger set (type).  A Union of sets (types) is also set (type).  It's legal to take an arbitrary subset (subtype) of any larger set (type).  Am I wrong there?

**Reply by erictraut** [reply]
> I had believed the role of NewType was to declare an arbitrary subset (subtype) of a larger set (type).

No, that's not the role of `NewType`. It's meant to be a stand-in (a proxy) for a specified class. At runtime, the NewType is erased completely, and it's the same as the class it is proxying. At static analysis time, the NewType is treated as a subclass of the specified class. This allows you to statically distinguish between specific uses of the class being proxied. For example, let's say that your function accepts two float values -- the first represents a "height" value and the second a "width" value. If you want to help the caller avoid confusing the two, you can do the following:

```python
Height = NewType("Height", float)
Width = NewType("Width", float)

def computeArea(h: Height, w: Width) -> float:
    return h * w

computeArea(1.0, 1.4) # Error
computeArea(Height(1.0), Width(1.4)) # Works
```

**Reply by Jacob-Stevens-Haas** [reply]
Oh, ok, that makes sense.  Thanks for clarifying, and thanks for sticking with me through this question. :)

**Comment by Jacob-Stevens-Haas** [comment]
Relatedly, I've also found a bit about `typing.Protocol`.  That comes close, because `scipy.sparse.spmatrix` and `numpy.ndarray` are the only classes I know that implement python's `@` operator via `__matmul__`.  However, Pyright also doesn't like using `NewType` around a `Protocol`.  If the `Intersection` types considered in PEP 483 were implemented in Python, I could achieve what I'm trying to with the intersection of a protocol for `__matmul__` with a `NewType(Foo, object)`.

**Reply by erictraut** [reply]
> Pyright also doesn't like using NewType around a `Protocol`

Protocols cannot be instantiated, so it wouldn't make sense to use them with `NewType`. There would be no way to instantiate the NewType because the constructor for a NewType requires you to pass an instance of the class. Not surprisingly, mypy also emits an error if you attempt to use `NewType` with a protocol, a union, or anything other than a normal (nominal) class.
